camera_alignment_core/alignment_utils/ring_alignment.py

```diff
@@ -44,30 +49,132 @@ def assign_ref_to_mov(
+                        mov_bead + 1
+                    ]
                 }
             )
-
         return ref_mov_coor_dict
 
+    def pos_bead_matches(
+        self,
+        ref_peak_dict: Dict[int, Tuple[int, int]],
+        mov_peak_dict: Dict[int, Tuple[int, int]],
+    ) -> Tuple[Dict[int, List[int]], Dict[int, int], Dict[int, int], int]:
+        """
+        Constrain ring matching problem by identifying which rings in the
+        moving image are closer to a given reference image ring than the
+        distance between rows/columns of rings. The threshold distance is
+        found by computing the median distance between rings and their 4
+        nearest neighbors in the reference image.
+
+        :param ref_peak_dict:  A dictionary
+            ({bead_number: (coor_y, coor_x)}) from reference beads
+        :param mov_peak_dict:  A dictionary
+            ({bead_number: (coor_y, coor_x)}) from moving beads
+        :return:
+            match_dict: A dictionary mapping the reference bead numbers
+                and moving bead numbers within the threshold distance
+            cost_dict: A dictionary mapping distance between a
+                reference bead numbers and matching moving beads
+            thresh_dist: The threshold distance calculated
+        """
+        # calculate threshold distance with mean distance to nearest neighbors
+        neigh = NearestNeighbors(n_neighbors=4)
+        neigh.fit([coors for coors in ref_peak_dict.values()])
+        mean_dist = []
+        for coor in ref_peak_dict.values():
+            dist, _ = neigh.kneighbors([coor], n_neighbors=4)
+            mean_dist.append(dist)
+
+        thresh_dist = np.median(mean_dist) * 0.9
+
+        offset = self.calc_cross_offset()
+        offset_mag = sqrt(offset[0] ** 2 + offset[1] ** 2)
+        if offset_mag > thresh_dist:
+            offset = (0, 0)
+
+        # generate bead neighborhood
+        tree_ref = KDTree(np.array([coors for coors in ref_peak_dict.values()]))
+        tree_mov = KDTree(np.array([coors for coors in mov_peak_dict.values()]))
+        neigh = tree_ref.query_ball_tree(tree_mov, thresh_dist)
+
+        # match reference beads to moving beads within threshold distance
+        match_dict = {}
+        cost_dict = {}
+        ref_beads = list(ref_peak_dict.keys())
+        mov_beads = list(mov_peak_dict.keys())
+        for idx_ref, idxs_mov in enumerate(neigh):
+            matches = []
+            costs = []
+            for idx_mov in idxs_mov:
+                matches.append(mov_beads[idx_mov])
+                dist = distance.euclidean(
+                    ref_peak_dict[ref_beads[idx_ref]],
+                    tuple(
+                        [
+                            m - o
+                            for m, o in zip(mov_peak_dict[mov_beads[idx_mov]], offset)
+                        ]
+                    ),
+                )
+                costs.append(dist + 0.001)
```

Comment:
What is the 0.001?  Something to prevent a zero value?

Reply:
yes, zero value prevention